management/src/main/java/io/micronaut/management/endpoint/loggers/LoggersEndpoint.java

```diff
@@ -0,0 +1,110 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.micronaut.management.endpoint.loggers;
+
+import io.micronaut.http.HttpResponse;
+import io.micronaut.http.annotation.QueryValue;
+import io.micronaut.http.hateos.JsonError;
+import io.micronaut.management.endpoint.Endpoint;
+import io.micronaut.management.endpoint.EndpointConfiguration;
+import io.micronaut.management.endpoint.Read;
+import io.micronaut.management.endpoint.Write;
+import io.reactivex.Single;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotBlank;
+
+/**
+ * Exposes an {@link Endpoint} to manage loggers.
+ *
+ * @author Matthew Moss
+ * @since 1.0
+ */
+@Endpoint(id = LoggersEndpoint.NAME,
+        defaultEnabled = LoggersEndpoint.DEFAULT_ENABLED,
+        defaultSensitive = LoggersEndpoint.DEFAULT_SENSITIVE)
+public class LoggersEndpoint {
+
+    /**
+     * Endpoint name.
+     */
+    public static final String NAME = "loggers";
+
+    /**
+     * Endpoint configuration prefix.
+     */
+    public static final String PREFIX = EndpointConfiguration.PREFIX + "." + NAME;
+
+    /**
+     * Endpoint default enabled.
+     */
+    public static final boolean DEFAULT_ENABLED = true;
+
+    /**
+     * Endpoint default sensitivity.
+     */
+    public static final boolean DEFAULT_SENSITIVE = false;
+
+    private final LoggingSystem loggingSystem;
+    private final LoggersManager loggersManager;
+
+    /**
+     * @param loggingSystem the {@link LoggingSystem}
+     * @param loggersManager the {@link LoggersManager}
+     */
+    public LoggersEndpoint(LoggingSystem loggingSystem,
+                           LoggersManager loggersManager) {
+        this.loggingSystem = loggingSystem;
+        this.loggersManager = loggersManager;
+    }
+
+    /**
+     * @return the loggers as a {@link Single}
+     */
+    @Read
+    public Single loggers() {
+        return Single.fromPublisher(loggersManager.getLoggers(loggingSystem));
+    }
+
+    /**
+     * @param name The name of the logger to find
+     * @return the {@link LogLevel} (both configured and effective) of the named logger
+     */
+    @Read
+    public Single logger(@QueryValue @NotBlank String name) {
+        return Single.fromPublisher(loggersManager.getLogger(loggingSystem, name));
+    }
+
+    /**
+     * @param name The name of the logger to configure
+     * @param configuredLevel The {@link LogLevel} to set on the named logger
+     * @return The {@link HttpResponse} with status code and message on error
+     */
+    @Write
+    public HttpResponse setLogLevel(@QueryValue @NotBlank String name,
```

Comment:
So what should the client expect if an error occurs... or is the goal to have endpoints not respond with errors?

The Spring returns a 400 and json (a JsonError or similar). Would returning an Optional<JsonError> be appropriate/correct?

Reply:
I'm not sure what the correct answer for this one is. I'm not sure if you are advocating that endpoints should not return errors, or not return an explicit `HttpResponse`.

If the former, then I need to remove things like `@NotBlank`? ... since those work as expected and respond with `BAD_REQUEST` and an appropriate message if `name` is blank. And then just make the function void and return nothing, and trap/consume all exceptions?

If the latter... letting an `IllegalArgumentException` through seems wrong. I could do something like this:
```
    try {
        // actual work here...
    } catch (IllegalArgumentException ex) {
        throw new HttpStatusException(HttpStatus.BAD_REQUEST, ex.getMessage());
    }
```
or does this still violate direct reference to HTTP concepts? Is there a more appropriate exception to throw?

The Spring loggers endpoint does generate 400 errors if you send a bad log level or try to set the root logger to null.

Reply:
Sorry that should have read `should not` use http concepts. No you don't need to remove `@NotBlank`, since that has nothing to do with HTTP

As for exception handling, leaving it throwing `HttpStatusException` for now (as described above), but we will have to introduce a common endpoint exception and native handler for it so it can be removed.